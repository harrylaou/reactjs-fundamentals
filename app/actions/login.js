export const LOGIN_USER_OK = 'LOGIN_USER_OK'
export const LOGOUT_USER = 'LOGOUT_USER'

export const loginUser = () => ({
  type: LOGIN_USER_OK
})

export const logoutUser = () => ({
  type: LOGOUT_USER
})