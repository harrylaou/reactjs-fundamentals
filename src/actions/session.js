export const LOG_OUT = 'LOG_OUT'
export const LOGGED_IN = 'LOGGED_IN'

export const logout = () => ({
  type: LOG_OUT
})

export const loggedIn = token => ({
  type: LOGGED_IN,
  token
})
