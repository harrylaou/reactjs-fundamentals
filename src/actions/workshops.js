export const RECEIVE_WORKSHOPS = 'RECEIVE_WORKSHOPS'

export const receiveWorkshops = workshops => ({
  type: RECEIVE_WORKSHOPS,
  workshops
})
