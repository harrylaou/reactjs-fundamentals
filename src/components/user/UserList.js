import React, { Component } from 'react'
import { List, ListItem } from 'material-ui/List'
import ActionGrade from 'material-ui/svg-icons/action/grade'
import Avatar from 'material-ui/Avatar'
import { pinkA200 } from 'material-ui/styles/colors'
import { LARGE } from '../../utils/WithWidth'

const UserList = (props) => {
  let listItems = props.users.map(user => (
    <ListItem
      onClick={props.showUserProfile.bind(this, user)} key={user.username} style={{color: "black"}}
      primaryText={ `${user.name.first} ${user.name.last}`}
      leftIcon={<ActionGrade color={pinkA200} />}
      rightAvatar={<Avatar src={`images/${user.username}_sm.jpg`} />}
    />
  ))

  const listView = <List className="view">{listItems}</List>
  const childrenView = <div className="view">{props.children}</div>

  let masterView, detailView

  if (props.username && props.width !== LARGE) {
    masterView = childrenView
    detailView = null
  } else {
    masterView = listView
    detailView = childrenView
  }
  return (
    <div className="view-container">
      {masterView}
      {detailView}
    </div>
  )
}

UserList.propTypes = {
  users: React.PropTypes.array.isRequired,
  username: React.PropTypes.string,
  width: React.PropTypes.number.isRequired
}

export default UserList
