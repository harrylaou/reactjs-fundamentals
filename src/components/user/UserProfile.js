import React, { Component } from 'react'
import {
  Card, CardActions, CardHeader,
  CardMedia, CardTitle
} from 'material-ui/Card'

const UserProfile = ({ user, username }) => {
  let fullname = ''
  let email = ''
  let gender = ''

  if (user){
    fullname = `${user.name.title} ${user.name.first} ${user.name.last}`
    email = user.email
    gender = user.gender
  }

  return (
    <Card>
      <CardHeader
        title={fullname}
        subtitle={username}
        avatar={`/images/${username}_sm.jpg`}
      />
      <CardMedia
        overlay={<CardTitle title={email} subtitle={gender} />}
      >
        <img src={`/images/${username}_lg.jpg`} />
      </CardMedia>
      <CardActions />
    </Card>
  )
}

UserProfile.propTypes = {
  user: React.PropTypes.object,
  username: React.PropTypes.string
}

export default UserProfile
