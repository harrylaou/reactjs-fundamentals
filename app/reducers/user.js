import {RECEIVE_USER} from '../actions/user'

const user = (state = null, action) => {
  switch (action.type) {
    case RECEIVE_USER: {
      return action.user
    }
    default:
      return state
  }
}

export default user