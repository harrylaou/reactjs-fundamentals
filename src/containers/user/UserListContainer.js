import React, { Component } from 'react'
import { withRouter } from 'react-router'
import { connect } from 'react-redux'
import UserList from '../../components/user/UserList'
import * as api from '../../api'
import * as actions from '../../actions/users'
import withWidth from '../../utils/WithWidth'

export class UserListContainer extends Component {
  constructor() {
    super()
    this.showUserProfile = this.showUserProfile.bind(this)
  }

  componentDidMount() {
    // Imagine we want to fetch users from another component. How do you think
    // we could reuse this code?
    api
      .getUsers()
      .then((users) => {
        this.props.dispatch(actions.receiveUsers(users))
      })
      .catch((err)=> {
        console.log(err)
      })

  }

  showUserProfile(user) {
    this.props.router.push(`/users/${user.username}`)
  }

  render() {
    return (
      <UserList
        {...this.props}
        username={this.props.params.username}
        users={this.props.users}
        showUserProfile={this.showUserProfile}
        width={this.props.width}
      />
    )
  }
}

UserListContainer.propTypes = {
  params: React.PropTypes.object.isRequired,
  width: React.PropTypes.number.isRequired,
  router: React.PropTypes.object.isRequired,
  users: React.PropTypes.array,
}

const mapStateToProps = (state) => ({
  users: state.users
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = (dispatch) => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(withWidth(UserListContainer)))
