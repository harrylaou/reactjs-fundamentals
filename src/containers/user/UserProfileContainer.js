import React, { Component } from 'react'
import { connect } from 'react-redux'
import UserProfile from '../../components/user/UserProfile'
import * as api from '../../api'
import * as actions from '../../actions/user'

export class UserProfileContainer extends Component {
  constructor() {
    super()
    this.fetchUser = this.fetchUser.bind(this)
  }

  componentDidMount() {
    this.fetchUser(this.props.params.username)
  }

  componentWillReceiveProps(nextProps) {
    // If you are thinking that performance could be improved by caching
    // the user, we agree. But we want to keep this example simple.
    // We are going to improve this code in the next lessons. We are building up :)
    if (this.props.params.username !== nextProps.params.username) {
      this.fetchUser(nextProps.params.username)
    }
  }

  // Imagine we want to fetch a user from another component. How do you think we
  // could reuse this code?
  fetchUser(username) {
    api
      .getUser(username)
      .then((user) => {
        this.props.dispatch(actions.receiveUser(user))
      })
      .catch((err)=> {
        console.log(err)
      })
  }

  render() {
    return (
      <UserProfile
        {...this.props}
        user={this.props.user}
        username={this.props.params.username}
      />
    )
  }
}

UserProfileContainer.propTypes = {
  params: React.PropTypes.object.isRequired,
  user: React.PropTypes.object
}

const mapStateToProps = (state) => ({
  user: state.user
})

const mapDispatchToProps = (dispatch) => ({
  dispatch
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserProfileContainer)