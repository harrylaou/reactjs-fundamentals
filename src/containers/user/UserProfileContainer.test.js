import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import ConnectedUserProfileContainer, {UserProfileContainer} from './UserProfileContainer'
import { mockFetch } from '../../testUtils'
import getMuiTheme from 'material-ui/styles/getMuiTheme'

import FakeConfigureStore from 'redux-mock-store'
import ActualConfigureStore from '../../store'
import {Provider} from 'react-redux'
import * as actions from '../../actions/user'


describe('<UserProfileContainer />', () => {
  it('should render with default props', () => {
    const props = {
      params: { username: '123' }
    }

    const wrapper = shallow(
      <UserProfileContainer {...props} />
    )

    expect(shallowToJson(wrapper.dive())).toMatchSnapshot()
  })


  const userData = { user : "Eduard" }
  const initialState = { user : "Manolo" }
  const mockStore = FakeConfigureStore()
  const props = {
    params: { username: '123' }
  }
  let store

  describe('<ConnectedUserProfileContainer />', () => {
    it('should render connected component', () => {
      store = mockStore(initialState)
      const container = shallow(
        <ConnectedUserProfileContainer
          {...props}
          store={store}
        />
      )

      expect(shallowToJson(container)).toMatchSnapshot()
    })
  })

  describe('Integration test: Mount + Wrapping Provider', () => {
    let wrapper

    beforeEach(()=>{
      store = mockStore(initialState)

      mockFetch(userData)

      wrapper = mount(
        <Provider store={store} >
          <ConnectedUserProfileContainer
            {...props}
          />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes:{
            muiTheme: React.PropTypes.object.isRequired,
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedUserProfileContainer).length).toEqual(1)
    })

    it('should Prop matches with initialState', () => {
      expect(wrapper.find(UserProfileContainer).prop('user')).toEqual(initialState.user)
    })

    it('should check action dispatch', () => {
      store.dispatch(actions.receiveUser(userData))
      const action = store.getActions()

      expect(action[0].type).toBe(actions.RECEIVE_USER)
    })
  })

  describe('Integration test: Mount + Wrapping Provider + Actual Store + Reducers', () => {
    let wrapper

    beforeEach(()=>{
      store = ActualConfigureStore()

      mockFetch(userData)

      wrapper = mount(
        <Provider store={store} >
          <ConnectedUserProfileContainer
            {...props}
          />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes:{
            muiTheme: React.PropTypes.object.isRequired,
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedUserProfileContainer).length).toEqual(1)
    })

    it('should check action dispatch', () => {
      //OR store.dispatch(actions.receiveUser(userData))
      wrapper.find(UserProfileContainer).props().dispatch(actions.receiveUser(userData))

      expect(wrapper.find(UserProfileContainer).prop('user')).toEqual(userData)
    })
  })
})