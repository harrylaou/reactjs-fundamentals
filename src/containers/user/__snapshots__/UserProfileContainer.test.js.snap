// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfileContainer /> <ConnectedUserProfileContainer /> should render connected component 1`] = `
<UserProfileContainer
  dispatch={[Function]}
  params={
    Object {
      "username": "123",
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
  user="Manolo"
/>
`;

exports[`<UserProfileContainer /> should render with default props 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
>
  <CardHeader
    avatar="/images/123_sm.jpg"
    subtitle="123"
    title=""
  />
  <CardMedia
    overlay={
      <CardTitle
        subtitle=""
        title=""
      />
    }
  >
    <img
      src="/images/123_lg.jpg"
    />
  </CardMedia>
  <CardActions />
</Card>
`;
